### PR TITLE
Append UUID string to the tmp dir to create unique dir

### PR DESCRIPTION
**Pull request by rvishureddy**
Create a uuid string append to the temp dir to create a unique temp directory.

### CODE REVIEW

File: cpp/FileTransferAgent.cpp [hunk 1]
```diff
@@ -300,7 +300,8 @@ void Snowflake::Client::FileTransferAgent::compressSourceFile(
 {
   CXX_LOG_DEBUG("Starting file compression");
   
-  char tempDir[100];
+  char tempDir[150]={0};
```
Comment (smtakeda) [inline cpp/FileTransferAgent.cpp]: can this be larger? 150 is too small for full path. could this be 4096?

File: lib/platform.c [hunk 2]
```diff
@@ -543,20 +546,24 @@ int STDCALL sf_delete_directory_if_exists(const char * directoryName)
...
-  }
+    sprintf(tmpDir,"/tmp/%s/",uuid_cstr);
   else
-  {
-    strncpy(tmpDir, tmpEnv, strlen(tmpEnv));
-    size_t oldLen = strlen(tmpDir);
-    tmpDir[oldLen] = PATH_SEP;
-    tmpDir[oldLen+1] = '\0';
-  }
+    sprintf(tmpDir, "%s/%s/",tmpEnv,uuid_cstr);
```
Comment (smtakeda) [inline lib/platform.c]: add curly braces even for online statement.

File: lib/platform.c [hunk 3]
```diff
@@ -543,20 +546,24 @@ int STDCALL sf_delete_directory_if_exists(const char * directoryName)
...
-
+  uuid_t uuid;
+  char uuid_cstr[37]; // 36 byte uuid plus null.
+  uuid_generate(uuid);
+  uuid_unparse(uuid, uuid_cstr);
   if (!tmpEnv)
-  {
-    strncpy(tmpDir, "/tmp/", sizeof("/tmp/"));
-  }
+    sprintf(tmpDir,"/tmp/%s/",uuid_cstr);
```
Comment (smtakeda) [inline lib/platform.c]: add curly braces.

File: lib/platform.c [hunk 4]
```diff
@@ -543,20 +546,24 @@ int STDCALL sf_delete_directory_if_exists(const char * directoryName)
...
-    strncpy(tmpDir, tmpEnv, strlen(tmpEnv));
-    size_t oldLen = strlen(tmpDir);
-    tmpDir[oldLen] = PATH_SEP;
-    tmpDir[oldLen+1] = '\0';
-  }
+    sprintf(tmpDir, "%s/%s/",tmpEnv,uuid_cstr);
 #endif
-}
\ No newline at end of file
+  sf_create_directory_if_not_exists(tmpDir);
```
Comment (smtakeda) [inline lib/platform.c]: If the directory exists, it is very likely the tmp directory is already used by other PUT commands?

File: cpp/FileTransferAgent.cpp (hunk 1 above)
Comment (rvishureddy) [reply cpp/FileTransferAgent.cpp]: Was adding extra 37 bytes of UUID so thought +50 should do. But yeah customers can have long path. Will fix it.

File: lib/platform.c (hunk 4 above)
Comment (rvishureddy) [reply lib/platform.c]: This is going to be a unique temp directory for every put command. Every time the customer tries to put file we create a unique temp directory. which looks like
/tmp/<uuid string>/

File: cpp/FileTransferAgent.cpp [hunk 5]
```diff
@@ -300,7 +300,8 @@ void Snowflake::Client::FileTransferAgent::compressSourceFile(
 {
   CXX_LOG_DEBUG("Starting file compression");
   
-  char tempDir[100];
+  char tempDir[4096]={0};
+  memset(tempDir, 0, 4096);
```
Comment (ChTimTsubasa) [inline cpp/FileTransferAgent.cpp]: The previous line should already set the array to 0 and there should be no need to memset again.

File: lib/platform.c [hunk 6]
```diff
@@ -542,21 +543,22 @@ int STDCALL sf_delete_directory_if_exists(const char * directoryName)
 
 void STDCALL sf_get_tmp_dir(char * tmpDir)
```
Comment (ChTimTsubasa) [inline lib/platform.c]: Can you add comments to the function declaration informing user the minimum memory required for tmpDir?

File: cpp/FileTransferAgent.cpp (hunk 5 above)
Comment (rvishureddy) [reply cpp/FileTransferAgent.cpp]: True. Removed the later.

File: lib/platform.c (hunk 6 above)
Comment (rvishureddy) [reply lib/platform.c]: Left the original funtion as is and created a new function sf_get_uniq_tmp_dir which has comments.

**Comment by rvishureddy** [comment]
> One suggestion is to have a common dir prefix under tmp instead of directly using uuid under tmp. For example it could be /tmp/snowflake_local_stage/{env}/{uuid}. That way it could be easier if we want to clean up later.
Its already temp dir it does not give any advantage of having one more local directory. Even if you have a local_stage you cannot remove with a wild char as there might be some folders being in use. And then can always be removed by sorting by date.

**Comment by notkriswagner** [comment]
> > One suggestion is to have a common dir prefix under tmp instead of directly using uuid under tmp. For example it could be /tmp/snowflake_local_stage/{env}/{uuid}. That way it could be easier if we want to clean up later.
> > Its already temp dir it does not give any advantage of having one more local directory. Even if you have a local_stage you cannot remove with a wild char as there might be some folders being in use. And then can always be removed by sorting by date.

One advantage of the common dir prefix is that if a customer looks at their /tmp dir then they would know that these temp dirs are created by libsnowflakeclient instead of wondering what the random UUIDs mean

**Pull request merged**